File: parciales/heap.py
def posicion_padre(posicion_hijo):
	if (posicion_hijo == 0):
		return -1;
	return (int)((posicion_hijo - 1) / 2)

def posicion_hijo_der(posicion_padre):
	return (int)((posicion_padre * 2) + 2)

def posicion_hijo_izq(posicion_padre):
	return (int)((posicion_padre * 2) + 1)

def swap_vector(arr,pos1,pos2):
	arr[pos1],arr[pos2] = arr[pos2],arr[pos1]

class Heap:
	def __init__(self,comparar):
		self.cantidad = 0
		self.comparar = comparar
		self.arreglo = []

	def dar_arreglo(self):
		"""debug"""
		return self.arreglo

	def upheap(self, pos):
		padre = posicion_padre(pos)
		if(padre == -1):
			return
		elif(self.comparar(self.arreglo[pos], self.arreglo[padre]) > 0):
			swap_vector(self.arreglo, pos, padre)
			self.upheap(padre)

	def downheap(self, pos):
		while(pos < self.cantidad):
			pos_max = pos
			izq = posicion_hijo_izq(pos)
			der = posicion_hijo_der(pos)
			if(izq < self.cantidad and self.comparar(self.arreglo[izq], self.arreglo[pos_max]) > 0):
				pos_max = izq
			if(der < self.cantidad and self.comparar(self.arreglo[der], self.arreglo[pos_max]) > 0):
				pos_max = der
			if(pos_max == pos):
				return
			swap_vector(self.arreglo, pos_max, pos)
			self.downheap(pos_max)
			pos = pos_max

	def heap_cantidad(self):
		return self.cantidad

	def heap_esta_vacio(self):
		return (self.cantidad == 0)

	def heap_encolar(self, elem):
		self.arreglo.append(elem)
		self.upheap(self.cantidad)
		self.cantidad += 1
		return True

	def heap_ver_max(self):
		if(self.heap_esta_vacio()):
			return None
		return self.arreglo[0]

	def heap_desencolar(self):
		if (self.cantidad == 0):
			return None
		self.cantidad -= 1
		swap_vector(self.arreglo,0,self.cantidad)
		devolver = self.arreglo[self.cantidad]
		self.arreglo[self.cantidad] = None
		self.arreglo = self.arreglo[:self.cantidad]
		self.downheap(0)
		return devolver

	def heap_crear_arr(arreglo, n, comp):
		heap_t* heap_nuevo = Heap(comp)
		for i in range(n):
			heap_nuevo->arreglo[i] = arreglo[i]
		heap_nuevo->cantidad = n
		heap_nuevo->comparar = comp
		for j in range(posicion_padre(n - 1),0,-1)
			downheap(heap_nuevo, (size_t)pos)
		return heap_nuevo

File: heap/heap.c
/* CONSTANTES */
const size_t TAM_BASE = 20;

struct heap{
	size_t cantidad;
	size_t cant_max;
	void* arreglo[];
	cmp_func_t comparar;//???
}

/*typedef int (*cmp_func_t) (const void *a, const void *b);

menor a 0 si a < b
0 si a == b
mayor a 0 si a > b

*/

/* cambia las pociciones de dos punteros en un arreglo */
void swap_vector(void* vector,size_t pos_1, size_t pos_2){
	void* algo = vector[pos_1];
	vector[pos_1] = vector[pos_2];
	vector[pos_2] = algo;
	return;
}

size_t posicion_padre(size_t posicion_hijo){
	if (posicion_hijo == 0){return NULL;}
	return ((posicion_hijo - 1) / 2)
}

size_t posicion_hijo_der(size_t posicion_padre){
	return ((posicion_padre * 2) + 2);
}

size_t posicion_hijo_izq(size_t posicion_padre){
	return ((posicion_padre * 2) + 1);
}

bool upheap(heap_t* heap,size_t pos){
}

bool downheap(heap_t* heap,size_t pos){
}

heap_t *heap_crear(cmp_func_t cmp){
	heap_t* heap_nuevo = malloc(sizeof(heap_t));
	if (heap_nuevo == NULL){return NULL;}
	heap_nuevo->arreglo = malloc(sizeof(void*) * TAM_BASE);
	if (heap_nuevo->arreglo == NULL){return NULL;}
	heap_nuevo->cantidad = 0;
	heap_nuevo->comparar = cmp;
	heap_nuevo->cant_max = TAM_BASE;
	return heap_nuevo;
}

heap_t *heap_crear_arr(void *arreglo[], size_t n, cmp_func_t cmp){
	heap_t* heap_nuevo = malloc(sizeof(heap_t));
	if (heap_nuevo == NULL){return NULL;}
	heap_nuevo->arreglo = arreglo;
	heap_nuevo->cantidad = n;
	heap_nuevo->cant_max = n;
	heap_nuevo->comparar = cmp;
	size_t pos = posicion_padre(n);//toma +- desde el ultimo "nodo" con al menos un hijo
	while (pos > 0){
		downheap(heap_nuevo,pos);
		pos--;
	}
	downheap(heap_nuevo->arreglo[0];
	return heap_nuevo;
}

void heap_destruir(heap_t *heap, void destruir_elemento(void *e)){
}

size_t heap_cantidad(const heap_t *heap){
	return heap->cantidad;
}

bool heap_esta_vacio(const heap_t *heap){
	return (heap->cantidad == 0);
}

bool heap_encolar(heap_t *heap, void *elem){
	heap->arreglo[heap->cantidad] = elem;
	upheap(heap,heap->cantidad);
	heap->cantidad++;
	return true;
}

void *heap_ver_max(const heap_t *heap){
	return heap->arreglo[0];
}

void *heap_desencolar(heap_t *heap){
	if (heap->cantidad == 0){return NULL;}
	hash->cantidad--;
	swap_vector(heap->arreglo,0,heap->cantidad);//ya se le resto uno antes
	void* devolver = heap->arreglo[heap->cantidad];
	heap->arreglo[heap->cantidad] = NULL;
	downheap(hash,0);
	return devolver;
}

/* FUNCION DE ORDENAMIENTO */
void heap_sort(void *elementos[], size_t cant, cmp_func_t cmp){
}
//////////////////////////////////////////////////
